src/device/wg_controller.rs

```diff
@@ -154,6 +167,103 @@ async fn consolidate_wg_fwmark<W: WireGuard>(
+    //        a decision to downgrade its connection - running this endpoint
+    //        invalidation just before WG peer consolidation would delay endpoint
+    //        invalidation until the next consolidation cycle. This is undesirable,
+    //        therefore endpoint invalidation checks should be run before *and*
+    //        after WireGuard peer consolidation.
+
+    //Retreive some helper values
+    let (proxy_endpoints, cpc, checked_endpoints) = match (proxy, cross_ping_check) {
+        (Some(proxy), Some(cpc)) => (
+            proxy.get_endpoint_map().await?,
+            cpc,
+            cpc.get_validated_endpoints().await?,
+        ),
+        _ => {
+            return Ok(()); //Meshnet is disabled
+        }
+    };
+    let actual_peers = wireguard_interface.get_interface().await?.peers;
+
+    // Each peer endpoint invalidation is processed individually
+    for (public_key, actual_peer) in actual_peers {
+        let is_proxying = is_peer_proxying(&actual_peer, &proxy_endpoints);
+        let actual_endpoint = actual_peer.endpoint;
+        let changed_at = actual_peer.endpoint_changed_at;
+        let published_endpoint = checked_endpoints.get(&public_key);
+        let published_at = published_endpoint.map(|endpoint| endpoint.changed_at);
+        let mut should_invalidate_endpoints = false;
+
+        // At debug level -> log every endpoint invalidation consideration. For info - only once a
+        // decision to invalidate has been taken
+        telio_log_debug!(
+            "Considering endpoint invalidation.\
+                         Decision info: {is_proxying:?} \
+                         {actual_endpoint:?} {changed_at:?} \
+                         {published_endpoint:?} {published_at:?}"
+        );
+
+        // If endpoint has never changed in wireguard or we have no published endpoints - there is
+        // nothing to invalidate.
+        let (changed_at, published_at) = match (changed_at, published_at) {
+            (Some(changed_at), Some(published_at)) => (changed_at, published_at),
+            _ => continue,
+        };
+
+        // Any endpoint published before a actual downgrade has occured is considered failed
+        if is_proxying && (published_at < changed_at.0) {
+            telio_log_info!(
+                "Invalidating published EPs for {public_key:?} after downgrade.\
+                            Decision info: {actual_endpoint:?} {changed_at:?} \
+                            {published_endpoint:?} {published_at:?}"
+            );
+            should_invalidate_endpoints = true;
+        }
+
+        // In direct connection if we have roamed from one direct EP to another - old checked EP is
+        // considered not valid anymore.
+        // Note, that endpoint change caused by libtelio writing new endpoint into WireGuard interface
+        // is *not* considered roaming.
+        if !is_proxying && changed_at.1 == UpdateReason::Pull && (published_at < changed_at.0) {
```

Comment:
I had this implemented with variables first, but wanted to add specific reason (roaming vs downgrade) in logs which triggers invalidation. And then it becomes harder to avoid if's :/

Reply:
oh, but I missed the "roaming" word itself :D

Reply:
Updated logging lines. Although only subtly

Reply:
missed that diff before :D